tools/decompress-assemblies/main.cs

```diff
@@ -0,0 +1,126 @@
+			return 1;
+		}
+
+		static bool UncompressDLL (Stream stream, string filePath)
+		{
+			string outputFile = $"uncompressed-{Path.GetFileName (filePath)}";
+			bool retVal = true;
+
+			Console.WriteLine ($"Processing {filePath}");
+			//
+			// LZ4 compressed assembly header format:
+			//   uint magic;                 // 0x5A4C4158; 'XALZ', little-endian
+			//   uint descriptor_index;      // Index into an internal assembly descriptor table
+			//   uint uncompressed_length;   // Size of assembly, uncompressed
+			//
+			using (var reader = new BinaryReader (stream)) {
+				uint magic = reader.ReadUInt32 ();
+				if (magic == CompressedDataMagic) {
+					reader.ReadUInt32 (); // descriptor index, ignore
+					uint decompressedLength = reader.ReadUInt32 ();
+
+					int inputLength = (int)(stream.Length - 12);
+					byte[] sourceBytes = bytePool.Rent (inputLength);
+					reader.Read (sourceBytes, 0, inputLength);
+
+					byte[] assemblyBytes = bytePool.Rent ((int)decompressedLength);
+					int decoded = LZ4Codec.Decode (sourceBytes, 0, inputLength, assemblyBytes, 0, (int)decompressedLength);
+					if (decoded != (int)decompressedLength) {
+						Console.Error.WriteLine ($"  Failed to decompress LZ4 data of {filePath} (decoded: {decoded})");
+						retVal = false;
+					} else {
+						using (var fs = File.Open (outputFile, FileMode.Create, FileAccess.Write)) {
+							fs.Write (assemblyBytes, 0, assemblyBytes.Length);
+							fs.Flush ();
+						}
+						Console.WriteLine ($"  uncompressed to: {outputFile}");
+					}
+
+					bytePool.Return (sourceBytes);
+					bytePool.Return (assemblyBytes);
+				} else {
+					Console.WriteLine ($"  assembly is not compressed");
+				}
+			}
+
+			return retVal;
+		}
+
+		static bool UncompressDLL (string filePath)
+		{
+			using (var fs = File.Open (filePath, FileMode.Open, FileAccess.Read)) {
+				return UncompressDLL (fs, filePath);
+			}
+		}
+
+		static bool UncompressFromAPK (string filePath)
+		{
+			using (ZipArchive apk = ZipArchive.Open (filePath, FileMode.Open)) {
+				foreach (ZipEntry entry in apk) {
+					if (!entry.FullName.StartsWith ("assemblies/", StringComparison.Ordinal)) {
+						continue;
+					}
+
+					if (!entry.FullName.EndsWith (".dll", StringComparison.Ordinal)) {
+						continue;
+					}
+
+					using (var stream = new MemoryStream ()) {
+						entry.Extract (stream);
+						stream.Seek (0, SeekOrigin.Begin);
+						UncompressDLL (stream, $"{filePath}!{entry.FullName}");
+					}
+				}
+			}
+
+			return true;
+		}
+
+		static int Main (string[] args)
+		{
+			if (args.Length == 0) {
+				return Usage ();
+			}
+
+			bool haveErrors = false;
+			foreach (string file in args) {
+				string ext = Path.GetExtension (file);
+				if (String.Compare (".dll", ext, StringComparison.OrdinalIgnoreCase) == 0) {
+					if (!UncompressDLL (file)) {
+						haveErrors = true;
+					}
+					continue;
+				}
+
+				if (String.Compare (".apk", ext, StringComparison.OrdinalIgnoreCase) == 0) {
```

Comment:
This can be handy to extend to handle `.aab` files as well.

Reply:
Support for AAB added :)